nixos/modules/services/system/hot-resize.nix

```
@@ -0,0 +1,140 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+
+let
+  cfg = config.services.hotResize;
+in
+{
+  options.services.hotResize = {
+    enable = lib.mkEnableOption "the hot resize service for filesystems";
+
+    devices = lib.mkOption {
+      type = lib.types.listOf (
+        lib.types.submodule {
+          options = {
+            device = lib.mkOption {
+              type = lib.types.str;
+              example = "/dev/sda1";
+              description = "Block device path to resize";
+            };
+
+            fsType = lib.mkOption {
+              type = lib.types.enum [
+                "ext4"
+                "xfs"
+                "btrfs"
+              ];
+              default = "ext4";
+              description = "Filesystem type (supported: ext4, xfs, btrfs)";
+            };
+
+            mountPoint = lib.mkOption {
+              type = lib.types.str;
+              example = "/";
+              description = "Mount point of the filesystem to resize";
+            };
+          };
+        }
+      );
+      default = [ ];
+      example = lib.literalExpression ''
+        [
+          {
+            device = "/dev/sda1";
+            fsType = "ext4";
+            mountPoint = "/";
+          }
+          {
+            device = "/dev/sdb1";
+            fsType = "xfs";
+            mountPoint = "/data";
+          }
+        ]
+      '';
+      description = "List of devices to monitor for resizing";
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    systemd.services.hotResize = {
+      description = "Hot resize service for filesystems";
+      wantedBy = [ "multi-user.target" ];
+
+      path = [
+        pkgs.parted
+        pkgs.cloud-utils
+        pkgs.coreutils
+        pkgs.util-linux
+        pkgs.e2fsprogs
+        pkgs.xfsprogs
+        pkgs.btrfs-progs
+      ];
+
+      serviceConfig = {
+        Type = "oneshot";
+        RemainAfterExit = false;
+        ExecStart = pkgs.writeShellScript "resize-fs" ''
+          set -euo pipefail
```

```suggestion
        ExecStart = lib.getExe (pkgs.writeShellApplication ...)
```
Since you're already using `-euo pipefail`, might as well add shellcheck to the mix.

---

It is no joke. `writeShellApplication` will allow you to remove some of the boilerplate, like the `set -euo pipefail` statement, and could even allow you to separate the bash part of the module into another file (so you can then use normal bash tools like the LSP or `shfmt`. It also adds a build-time check for shellcheck compliance, which is easier and more maintainable than doing it yourself every time.

[Here's an example](https://github.com/NixOS/nixpkgs/blob/d74689ac642e8e35522030e54a64d44ed2391ab9/nixos/modules/virtualisation/xen-dom0.nix#L42-L69) on how this could be achieved, and the [accompanying bash script](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/virtualisation/xen-boot-builder.sh). See the Nixpkgs manual for more usage information.

See also the [relevant section in the Nixpkgs Manual](https://nixos.org/manual/nixpkgs/unstable/#trivial-builder-writeShellApplication).